torch_em/shallow2deep/prepare_shallow2deep.py

```diff
@@ -504,11 +509,94 @@ def random_points(
     return features[samples], labels[samples]
 
 
+def worst_tiles(
+    features, labels, rf_id,
+    forests, forests_per_stage,
+    sample_fraction_per_stage,
+    sampling_kwargs,
+    accumulate_samples=True,
+):
+    # check inputs
+    img_shape = sampling_kwargs.get('img_shape', None)
+    assert len(img_shape) in [2, 3], img_shape
+    tiles_shape = sampling_kwargs.get('tiles_shape', None)
+    assert len(tiles_shape) in [2, 3], tiles_shape
+
+    # get the corresponding random forest from the last stage
+    # and predict with it
+    last_forest = forests[rf_id - forests_per_stage]
+    pred = last_forest.predict_proba(features)
+
+    # labels to one-hot encoding
+    unique, inverse = np.unique(labels, return_inverse=True)
+    onehot = np.eye(unique.shape[0])[inverse]
+
+    # compute the difference between labels and prediction
+    diff = np.abs(onehot - pred).sum(axis=1)
+    assert len(diff) == len(features)
+
+    # reshape diff to image shape and apply convolution with 1-kernel
+    diff_img = torch.Tensor(diff.reshape(img_shape)[None, None])
+    kernel = torch.Tensor(np.ones(tiles_shape)[None, None])
+    diff_img_smooth = torch.nn.functional.conv2d(diff_img, weight=kernel) if len(img_shape)==2 \
+        else torch.nn.functional.conv3d(diff_img, weight=kernel)
+    diff_img_smooth = diff_img_smooth.detach().numpy().squeeze()
+    diff_img_shape = diff_img_smooth.shape
+    diff_img_smooth = diff_img_smooth.flatten()
+
+    # define lambda functions for better readability
+    # need to add tiles_shape[i] // 2 to get indices in original img
+    start = lambda i: 0 # -(tiles_shape[i] // 2)
+    end = lambda i: tiles_shape[i] # tiles_shape[i] // 2 + 1
+
+    # get training samples based on tiles around maxima
+    # of the label-prediction diff
+    nc = len(np.unique(labels))
+    n_samples_class = int(sample_fraction_per_stage * len(features)) // nc
+
+    # sample in a class balanced way
+    samples_per_class = [[]]*nc
+    indices_per_class = [np.where(labels == class_id)[0] for class_id in range(nc)]
+
+    # get maxima of the label-prediction diff
+    max_centers = np.argsort(diff_img_smooth)[::-1]
```

Comment:
Wouldn't we rather use a function that finds local maxima here? I think that would make more sense, because otherwise we can end up oversampling the global maximum.
There's https://scikit-image.org/docs/dev/api/skimage.feature.html#skimage.feature.peak_local_max or `vigra.analysis.localMaxima` you can use for this.

Reply:
Yeah, you're totally right. I could even add the `min_distance` parameter. Then I could make sure that there is no overlap of the tiles at all. Maybe then I can do the entire thing more efficient. I'll have a look into that :)